remoulade/brokers/rabbitmq.py

```diff
@@ -413,3 +407,84 @@ def ack(self):
 
     def nack(self, requeue):
         self._rabbitmq_message.nack(requeue)
+
+
+class ChannelPool:
+    """A pool of channels that can be used by the RabbitmqBroker.
+
+    The pool uses a synchronized queue as a backend, making sure that two threads never end up sharing a channel.
+
+    The channels are created lazily as the reservation requests comes.
+
+    The ChannelPool should be used via the `acquire` context manager, to make sure that a used channel is properly put
+    back into the pool.
+
+    Examples:
+
+      >>> channel_pool = ChannelPool(channel_factory=lambda connection: connection.channel(), pool_size=5)
+      >>> with channel_pool.acquire() as channel:
+      ...     channel.basic.publish(...)
+
+    Parameters:
+      channel_factory(function): Function that will be called to create new channels.
+      pool_size(int): The max size of the pool.
+
+    """
+    def __init__(self, channel_factory, *, pool_size):
+        self._channel_factory = channel_factory
+        self._pool = Queue(pool_size)
+        self._pool_size = pool_size
+
+        for _ in range(pool_size):
+            # The goal is to create lazily the channels, so Nones are put as stand-in for the channel that can be used
+            self.put(None)
+
+    @contextmanager
+    def acquire(self, timeout=None):
+        """
+        Parameters:
+          timeout(int): The max number of second to wait when fetching a channel from the pool. If None, it will wait
+            indefinitely to get a channel. Default None.
+
+        Raises:
+          ChannelPoolTimeout: when the timeout for reserving a channel is run out.
+        """
+        channel = self.get(timeout=timeout)
+        if channel is None:
+            channel = self._channel_factory()
+        try:
+            yield channel
+        finally:
+            self.put(channel)
+
+    def get(self, *, timeout=None):
+        try:
+            return self._pool.get(timeout=timeout)
+        except Empty:
+            raise ChannelPoolTimeout("Could not get any channel from the pool")
+
+    def put(self, channel):
+        try:
+            return self._pool.put_nowait(channel)
+        except Full:
+            pass
```

Comment:
this should not happen right ? why Queue(pool_size) ? and not just Queue() ?

Reply:
It can, in the very border case: thread1 acquire a channel, then thread2 raise a connection error => it gets the lock and clear the pool which goes back to `pool_size`, and then thread1 put back its channel, you would end up with a pool size of `pool_size + 1`